airflow/jobs.py

```diff
@@ -930,7 +932,20 @@ def _process_task_instances(self, dag, queue, session=None):
             run.dag = dag
             # todo: preferably the integrity check happens at dag collection time
             run.verify_integrity(session=session)
-            run.update_state(session=session)
+            finished_tasks = (
+                session
+                .query(TI.task_id, TI.state
+                       )
+                .filter(
+                    TI.dag_id == run.dag_id,
+                    TI.execution_date == run.execution_date,
+                    TI.state.in_([
+                        State.SUCCESS, State.FAILED,
+                        State.UPSTREAM_FAILED, State.SKIPPED])
```

Comment:
This  should probably be `State.finished()`. But that list doesn't include UPSTREAM_FAILED, and I don't know what adding it would do.  Could you take a look and try to work that out?

Reply:
It seems that it should include also UPSTREAM_FAILED, the behaviour dealing with finished tasks is quiet similar in most cases, I will run the tests and we will see :)